Improve railties' tests

bogdanvlviv opened the pull request:
- Remove redundant execution of `Dir.chdir(app_path) { }` in railties' tests
- <strike>Improve RakeTest#test_not_protected_when_previous_migration_was_not_production</strike>
  - <strike>Allow failure for `db:test:prepare` to ensure the test woun't raise error:
     `RuntimeError: rails command failed (1): bin/rails db:test:prepare 2>&1`</strike>
  - <strike>Prevent redundant execution of `bin/rails test` inside the test</strike>
</strike>

- Invoke rails command inside the railties' test app with TestHelpers::Generation#rails
  See #30520
- Improve RakeTest#test_db_test_prepare_when_using_sql_format
  - Remove redundant setting `RAILS_ENV` for `db:test:prepare`.
    `db:test:prepare` doesn't require it.

/cc @matthewd

matthewd reviewed (changes requested):
There are several very different changes going on in a single commit here, and instead of describing any of them, the title of the commit is just copied from the other PR.

bogdanvlviv committed:
Remove redundant execution of `Dir.chdir(app_path) { }` in railties' tests

bogdanvlviv committed:
Invoke rails command inside the railties' test app with TestHelpers::Generation#rails

See #30520

bogdanvlviv committed:
Improve RakeTest#test_db_test_prepare_when_using_sql_format

  - Remove redundant setting `RAILS_ENV` for `db:test:prepare`.
    `db:test:prepare` doesn't require it.

bogdanvlviv commented:
@matthewd Thanks for the correcting, I just split up changes to commits with explanations.

matthewd commented on railties/test/application/rake_test.rb:
```diff
@@ -40,7 +40,7 @@ def test_not_protected_when_previous_migration_was_not_production
         rails "generate", "model", "product", "name:string"
         rails "db:create", "db:migrate"
-        output = Dir.chdir(app_path) { rails("db:test:prepare", "test") }
+        output = rails("db:test:prepare", allow_failure: true)
```
Is this currently raising? What failure do we need to allow? :confused:

bogdanvlviv replied:
If for some reason `db:test:prepare` fails, the code below
```
refute_match(/ActiveRecord::ProtectedEnvironmentError/, output)
```
 won't be executed

instead we will see error
````
RuntimeError: rails command failed (1): bin/rails db:test:prepare 2>&1
````

matthewd replied:
Why would that be a bad thing?

If the command fails, I want the test to fail, not pass just because the failure doesn't contain that string.

bogdanvlviv replied:
Because if leave it as it is and `db:test:prepare` fails for some reason(for example: adding bad changes in a PR)

this line will never be executed
```
refute_match(/ActiveRecord::ProtectedEnvironmentError/, output)
```
so if `db:test:prepare` raise `ActiveRecord::ProtectedEnvironmentError` or any antother error, test status will be as `error` but not as `failure` or `passed`.

In this way, this test doesn't check what we expect.

matthewd replied:
If there is an error in that command, then an error has occurred, and so 'error' would be the correct result for the test.

bogdanvlviv replied:
@matthewd I just removed the changes

matthewd commented on railties/test/application/rake_test.rb:
```diff
@@ -40,7 +40,7 @@ def test_not_protected_when_previous_migration_was_not_production
         rails "generate", "model", "product", "name:string"
         rails "db:create", "db:migrate"
-        output = Dir.chdir(app_path) { rails("db:test:prepare", "test") }
+        output = rails("db:test:prepare")
```
Again, I think this is important to ensure `test` actually works, because our only assertion is a loose non-match.

It's really not good to make changes to tests just because you don't think something needs to be there.

bogdanvlviv replied:
@matthewd do we have to return back execution of `bin/rails test` that was removed by https://github.com/rails/rails/pull/30520/files#diff-e1c35224a4df7f73d90a9a90e45aab2bL32?

matthewd replied:
No. That one is never reached, because the prepare fails. And the following assertion is positive ("this must be there"), not negative ("this specific string should not be").

bogdanvlviv replied:
I deleted the changes

bogdanvlviv commented:
@matthewd This PR is ready for review.

matthewd commented on railties/test/application/rake_test.rb:
```diff
@@ -113,10 +113,8 @@ def test_should_not_eager_load_model_for_rake
-        assert system("bin/rails do_nothing RAILS_ENV=production"),
-               "should not be pre-required for rake even eager_load=true"
-      end
+      assert rails("do_nothing" "RAILS_ENV=production"),
```
Asserting on a value that's always true isn't very useful.

.. and this is missing a comma, so it's also not running the command it intends to.

bogdanvlviv replied:
@matthewd Should I just leave `rails("do_nothing", "RAILS_ENV=production")` without `assert`?

matthewd replied:
Hmm.. maybe we should make `do_nothing` do a puts, then assert on that. That way it would've caught the missing comma.

bogdanvlviv replied:
> Hmm.. maybe we should make do_nothing do a puts, then assert on that. That way it would've caught the missing comma.

Good idea! Done.
